resolving/npm-resolver/src/index.ts

```diff
@@ -244,6 +254,64 @@ async function resolveNpm (
   }
 }
 
+async function tryResolveFromJsr (
+  ctx: ResolveFromNpmContext,
+  wantedDependency: WantedDependency,
+  opts: Omit<ResolveFromNpmOptions, 'registry'>,
+  defaultTag: string
+): Promise<ResolveResult | null> {
+  if (!wantedDependency.pref?.startsWith('jsr:')) return null
+
+  const registry = ctx.authConfig?.['@jsr:registry']
+  if (!registry) {
+    // Q: Why throw an error instead of falling back to a default URL?
+    // A: The packages from JSR registry may have dependencies in the `@jsr` scope which would
+    //    cause confusing error messages should it not be defined.
+
+    throw new PnpmError('JSR_REGISTRY_NOT_DEFINED', 'Cannot use JSR registry without defining a registry for the `@jsr` scope', {
+      hint: 'Fix it by adding `@jsr:registry=https://npm.jsr.io` to .npmrc',
+    })
+  }
+
+  const pref = wantedDependency.pref.slice('jsr:'.length)
+  const spec = parseJsrPref(pref, wantedDependency.alias, defaultTag, registry)
+  if (!spec.name.startsWith('@')) {
+    throw new PnpmError('MISSING_JSR_PACKAGE_SCOPE', 'Package names from JSR must have scopes')
+  }
+  const jsrNameSuffix = spec.name.replace('@', '').replace('/', '__') // not replaceAll because we only replace the first of each character
+  spec.name = `@jsr/${jsrNameSuffix}`
```

Comment:
why is this whole code block not inside parseJsrPref?

Reply:
https://github.com/pnpm/pnpm/pull/9358/commits/0833d43dc8d4d49ea1b488aadebb5761ffebeeab